src/vmm/src/devices/virtio/vhost_user.rs

```diff
@@ -257,3 +464,517 @@ impl VhostUserHandle {
+        vuh.vu.protocol_features = backend_protocol_features;
+
+        let (acked_features, acked_protocol_features) = vuh
+            .negotiate_features(avail_features.bits(), avail_protocol_features)
+            .unwrap();
+        assert_eq!(acked_features, avail_features.bits());
+        assert_eq!(acked_protocol_features, avail_protocol_features.bits());
+        assert_eq!(vuh.vu.protocol_features, avail_protocol_features);
+        assert_eq!(
+            unsafe { *vuh.vu.hdr_flags.get() },
+            VhostUserHeaderFlag::empty()
+        );
+
+        // If PROTOCOL_FEATURES and REPLY_ACK are negotiated
+        let avail_features = VhostUserVirtioFeatures::all();
+        // Pretend backend has same features as frontend
+        vuh.vu.features = avail_features.bits();
+
+        let mut avail_protocol_features = VhostUserProtocolFeatures::empty();
+        avail_protocol_features.set(VhostUserProtocolFeatures::REPLY_ACK, true);
+
+        // Pretend backend has same features as frontend
+        vuh.vu.protocol_features = avail_protocol_features;
+
+        let (acked_features, acked_protocol_features) = vuh
+            .negotiate_features(avail_features.bits(), avail_protocol_features)
+            .unwrap();
+        assert_eq!(acked_features, avail_features.bits());
+        assert_eq!(acked_protocol_features, avail_protocol_features.bits());
+        assert_eq!(vuh.vu.protocol_features, avail_protocol_features);
+        assert_eq!(
+            unsafe { *vuh.vu.hdr_flags.get() },
+            VhostUserHeaderFlag::NEED_REPLY,
+        );
+    }
+
+    #[test]
+    fn test_update_mem_table() {
+        struct MockMaster {
+            regions: std::cell::UnsafeCell<Vec<VhostUserMemoryRegionInfo>>,
+        }
+
+        impl VhostUserHandleBackend for MockMaster {
+            fn set_mem_table(
+                &self,
+                regions: &[VhostUserMemoryRegionInfo],
+            ) -> Result<(), vhost::Error> {
+                unsafe { (*self.regions.get()).extend_from_slice(regions) }
+                Ok(())
+            }
+        }
+
+        let vuh = VhostUserHandleImpl {
+            vu: MockMaster {
+                regions: std::cell::UnsafeCell::new(vec![]),
+            },
+            socket_path: "".to_string(),
+        };
+
+        let region_size = 0x10000;
+        let file = TempFile::new().unwrap().into_file();
+        let file_size = 2 * region_size;
+        file.set_len(file_size as u64).unwrap();
+        let regions = vec![
+            (
+                FileOffset::new(file.try_clone().unwrap(), 0x0),
+                GuestAddress(0x0),
+                region_size,
+            ),
+            (
+                FileOffset::new(file.try_clone().unwrap(), 0x10000),
+                GuestAddress(0x10000),
+                region_size,
+            ),
+        ];
+
+        let guest_memory = GuestMemoryMmap::from_raw_regions_file(regions, false, false).unwrap();
```

Comment:
it looks like there is an uncovered case where a file offset None. Can it be covered too?

Reply:
If file offsets are `None`, `from_raw_regions_file` will fail, so no additional testing is needed.

Reply:
oh, so that branch is practically unreachable?

Reply:
In our `memfd` backed scenario yes. I can add a test that creates non file backed memory and triggers error in the `update_mem_table` method though, just in case.